driver-core/src/main/com/mongodb/internal/async/function/RetryState.java

```diff
@@ -48,25 +50,56 @@ public final class RetryState {
 
     private final LoopState loopState;
     private final int attempts;
-    @Nullable
-    private Throwable exception;
+    private final TimeoutContext timeoutContext;
+    private Throwable previouslyChosenException;
 
     /**
-     * @param retries A non-negative number of allowed retries. {@link Integer#MAX_VALUE} is a special value interpreted as being unlimited.
+     * Creates a {@code RetryState} with a positive number of allowed retries. {@link Integer#MAX_VALUE} is a special value interpreted as being unlimited.
+     * If {@link TimeoutContext#hasTimeoutMS()} is true, the number of attempts is unbounded until {@link TimeoutContext#hasExpired()} is true.
+     * Otherwise,the retry attempts count will be used to determine the maximum number of attempts.
+     * <p>
+     * It is possible to provide an additional {@code retryPredicate} in the {@link #doAdvanceOrThrow(Throwable, BiFunction, BiPredicate, boolean)} method,
+     * which can be used to stop retrying based on a custom condition additionally to {@code retires} and {@link TimeoutContext}.
+     * </p>
+     *
+     * @param retries A positive number of allowed retries. {@link Integer#MAX_VALUE} is a special value interpreted as being unlimited.
+     * @param timeoutContext A timeout context that will be used to determine if the operation has timed out.
      * @see #attempts()
      */
-    public RetryState(final int retries) {
-        assertTrue(retries >= 0);
-        loopState = new LoopState();
-        attempts = retries == INFINITE_ATTEMPTS ? INFINITE_ATTEMPTS : retries + 1;
+    public static RetryState withRetryableState(final int retries, final TimeoutContext timeoutContext) {
```

Comment:
Previously, we used to pass `0` to the constructor to indicate that we didn't want retries to occur. For example:

`return new RetryState(retry ? RetryState.RETRIES : 0);`

However, with the two separate methods for code readability, `withNonRetryableState()` and `withRetryableState(final int retries, final TimeoutContext timeoutContext)`, we can no longer pass `0` to `withRetryableState` because `0` now signifies "no retries.